### PR TITLE
Left join initial nested loop implementation

### PR DESCRIPTION
**Pull request by jussisaurio**
Left join algorithm, always using a nested loop join (reverse engineered from `sqlite3` with `pragma automatic_index = 0;` - thanks @pereman2 for the tip):

```
addr  opcode             p1    p2    p3    p4             p5  comment
----  -----------------  ----  ----  ----  -------------  --  -------
0     Init               0     25    0                    0   Start at 25
1     OpenReadAsync      0     2     0                    0   root=2
2     OpenReadAwait      0     0     0                    0   
3     OpenReadAsync      1     3     0                    0   root=3
4     OpenReadAwait      0     0     0                    0   
5     RewindAsync        0     0     0                    0   
6     RewindAwait        0     24    0                    0   
7       Integer          0     1     0                    0   r[1]=0 <-------- SET "left join match flag" TO FALSE
8       RewindAsync      1     0     0                    0   
9       RewindAwait      1     19    0                    0   
10        RowId          0     2     0                    0   r[2]=users.rowid
11        RowId          1     3     0                    0   r[3]=products.rowid
12        Ne             2     3     17                   0   if r[2]!=r[3] goto 17 <----- IF NO MATCH ON JOIN COL, JUMP TO 17 (NEXT RIGHT TABLE ROW)
13        Integer        1     1     0                    0   r[1]=1 <-- SET "left join match flag" TO TRUE
14        Column         0     1     4                    0   r[4]=users.first_name
15        Column         1     1     5                    0   r[5]=products.name
16        ResultRow      4     2     0                    0   output=r[4..5]
17      NextAsync        1     0     0                    0   
18      NextAwait        1     9     0                    0   
19      IfPos            1     22    0                    0   r[1]>0 -> r[1]-=0, goto 22 <-- IF MATCH FLAG IS TRUE, JUMP TO 22 (NEXT LEFT TABLE ROW)
20      NullRow          1     0     0                    0   Set cursor 1 to a (pseudo) NULL row <- IF MATCH FLAG IS FALSE, SET A "NULL FLAG" ON THE RIGHT TABLE CURSOR TO MAKE Insn::Column ALWAYS RETURN NULL
21      Goto             0     13    0                    0   JUMP BACK TO "Set flag to true", READ THE COLUMNS AGAIN, ALL RIGHT TABLE COLUMNS WILL BE NULLED
22    NextAsync          0     0     0                    0   
23    NextAwait          0     6     0                    0   
24    Halt               0     0     0                    0   
25    Transaction        0     0     0                    0   
26    Goto               0     1     0                    0 
```

Supports:

- foo LEFT JOIN bar on xxx
- foo LEFT JOIN bar WHERE xxx (no join condition)
- foo LEFT JOIN BAR on xxx WHERE yyy (both)

Architectural changes:

- Before instructions for handling WHERE/JOIN are emitted ('translated'), they are first introspected in `evaluate_conditions()`
- `evaluate_conditions()` returns a single `QueryConstraint`, which is an enum consisting of members `QueryConstraint::Left` and `QueryConstraint::Inner`. Both `QueryConstraint` variants can contain a `where_clause` property and a `join_clause` property. The reason I wrote it this way is that even without a join condition, you can still have a LEFT JOIN, e.g. `select * from foo left join bar WHERE foo.id = bar.id`. Even a WHERE clause without a join is a `QueryConstraint::Inner`, which is a bit confusing and should probably be renamed. 
- Adds pseudo `null_flag` to Cursors to facilitate nulling out columns in a left join with no match

Sorry, this PR got pretty massive but I wanted to cover quite a few test cases and new bugs kept popping up :D

### CODE REVIEW

**Comment by jussisaurio** [comment]
The main issue I have with my current approach here is: after parsing, aren't where clauses and join constraints basically equivalent? Should we try to unify their handling even more? E.g.

`select * from foo f left join bar b on f.id = b.id or f.name = b.name`
vs
`select * from foo f left join bar b where f.id = b.id or f_name = b.name`

shouldn't these be treated the same way? in fact, the second example currently doesn't work properly (the current code only considers something a JOIN when there is a constraint...)

EDIT: I think I'm going to try to unify these in the code...

**Comment by jussisaurio** [comment]
Tried to unify the handling of JOIN/WHERE, should now also e.g. supports LEFT JOIN without an ON condition but with a WHERE instead. Changed the PR description to reflect this

**Comment by penberg** [comment]
Hey @jussisaurio, sorry for the conflict in `all.test`. It should be trivial to fix I hope.

**Comment by penberg** [comment]
@jussisaurio Can you please open a ticket about adding support for `automatic_index` for tracking?

**Comment by jussisaurio** [comment]
Future stuff to think about, TLDR: sqlite makes all join conditions part of the where clause; uses special tags in the internal AST for conditions that came from an outer join.

```
Prior to analysis, the following transformations are made to shift all join constraints into the WHERE clause:

All NATURAL joins are converted into joins with a USING clause.
All USING clauses (including ones created by the previous step) are converted into equivalent ON clauses.
All ON clauses (include ones created by the previous step) are added as new conjuncts (AND-connected terms) in the WHERE clause.

SQLite makes no distinction between join constraints that occur in the WHERE clause and constraints in the ON clause of an inner join,
since that distinction does not affect the outcome. However, there is a difference between ON clause constraints and WHERE clause constraints for outer joins.
Therefore, when SQLite moves an ON clause constraint from an outer join over to the WHERE clause it adds special tags to the Abstract Syntax Tree (AST)
to indicate that the constraint came from an outer join and from which outer join it came.
There is no way to add those tags in pure SQL text. Hence, the SQL input must use ON clauses on outer joins.
But in the internal AST, all constraints are part of the WHERE clause, because having everything in one place simplifies processing.

After all constraints have been shifted into the WHERE clause, The WHERE clause is broken up into conjuncts (hereafter called "terms").
In other words, the WHERE clause is broken up into pieces separated from the others by an AND operator.
If the WHERE clause is composed of constraints separated by the OR operator (disjuncts) then the entire clause is considered to be a single "term"
to which the [OR-clause optimization](https://sqlite.org/optoverview.html#or_opt) is applied.
```
https://sqlite.org/optoverview.html

No reason why work on this couldn't start fairly early already to lay a better foundation. After this I could start looking ito fully rewriting all joins as part of a `Where` clause (i.e. binary ANDs). Just need to figure out a good way to attach metadata to the AST nodes like Sqlite is doing if there's outer joins involved.

File: core/translate.rs [hunk 1]
```diff
@@ -313,86 +313,503 @@ fn translate_where(select: &Select, program: &mut ProgramBuilder) -> Result<Opti
...
+                constraint_expr: where_clause.clone(),
+                no_match_jump_label: program.allocate_label(),
+                no_match_target_cursor: get_no_match_target_cursor(program, select, where_clause),
+            };
+            match op {
+                JoinOperator::TypedJoin { natural, join_type } => {
+                    if *natural {
+                        todo!("Natural join not supported");
+                    }
+                    // default to inner join when no join type is specified
```
Comment (pereman2) [inline core/translate.rs:395]: Looks like having all these conditions is too complex as there is duplicated code? I imagine you can have:
```
if where {
   do stuff with where
}
if join {
  do stuff with join
}
```
might be wrong tho

File: core/translate.rs [hunk 2]
```diff
@@ -313,86 +313,503 @@ fn translate_where(select: &Select, program: &mut ProgramBuilder) -> Result<Opti
...
+                                    constraint_expr: expr.clone(),
+                                    no_match_jump_label: left_join_no_match_label,
+                                    no_match_target_cursor: get_no_match_target_cursor(
+                                        program, select, expr,
+                                    ),
+                                }),
+                                found_match_next_row_label: left_join_found_match_next_row_label,
+                                match_flag: left_join_match_flag,
+                                match_flag_hit_marker: left_join_match_flag_hit_marker,
+                                left_cursor: HARDCODED_CURSOR_LEFT_TABLE, // FIXME: hardcoded
```
Comment (pereman2) [inline core/translate.rs:430]: great name xd

File: core/translate.rs [hunk 3]
```diff
@@ -313,86 +313,503 @@ fn translate_where(select: &Select, program: &mut ProgramBuilder) -> Result<Opti
...
+    };
+
+    Ok(constraint_maybe)
+}
+
+fn translate_conditions(
+    program: &mut ProgramBuilder,
+    select: &Select,
+    conditions: Box<Option<QueryConstraint>>,
+) -> Result<Box<Option<QueryConstraint>>> {
```
[the remark this answers is not in the document]
Comment (jussisaurio) [reply core/translate.rs:587]: No it isn't apparently. I couldn't get the compiler to agree with me earlier but I was probably just doing something weird. Will remove

[the remark this answers is not in the document]
Comment (jussisaurio) [reply core/translate.rs]: removed box

File: core/translate.rs (hunk 1 above)
Comment (jussisaurio) [reply core/translate.rs]: less repetition now
